errorcodes/errorcodes.go

```diff
@@ -0,0 +1,241 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package errorcodes
+
+import (
+	"errors"
+
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/runtime/protoiface"
+)
+
+const (
+	ErrorCodesFeatureMetadataKey = "error_codes_feature"
+	Owner                        = "components-contrib"
+	Domain                       = "dapr.io"
+	NoReason                     = "NO_REASON_SPECIFIED"
+)
+
+var NoErrorReason = WithErrorReason(NoReason, 503, codes.Internal)
+
+type ResourceInfo struct {
+	ResourceType string
+	ResourceName string
+}
+
+func WithErrorReason(reason string, httpCode int, grpcStatusCode codes.Code) ErrorOption {
+	f := func(er *DaprError) {
+		er.reason = reason
+		er.grpcStatusCode = grpcStatusCode
+		er.httpCode = httpCode
+	}
+	return f
+}
+
+// call this function to apply option
+type ErrorOption func(*DaprError)
```

Comment:
type/function declaration comments should ideally start with the name of the type or function name, i.e. `// ErrorOption applies a `DaprError` to..`